File: src/main/org/nlogo/gl/view/ViewControlToolBar.java
package org.nlogo.gl.view;

import org.nlogo.api.Agent;
import org.nlogo.api.I18N;
import org.nlogo.api.Perspective;

strictfp class ViewControlToolBar
    extends javax.swing.JToolBar {

  private final MouseMotionHandler inputHandler;

  private final javax.swing.JTextField status;
  private final javax.swing.JToggleButton orbitButton;
  private final javax.swing.JToggleButton zoomButton;
  private final javax.swing.JToggleButton moveButton;
  private final javax.swing.JToggleButton interactButton;

  private final javax.swing.AbstractAction ORBIT_ACTION =
      new MovementAction(I18N.gui().get("view.3d.orbit"), View.Mode.ORBIT);
  private final javax.swing.AbstractAction ZOOM_ACTION =
      new MovementAction(I18N.gui().get("view.3d.zoom"), View.Mode.ZOOM);
  private final javax.swing.AbstractAction MOVE_ACTION =
      new MovementAction(I18N.gui().get("view.3d.move"), View.Mode.TRANSLATE);
  private final javax.swing.AbstractAction INTERACT_ACTION =
      new MovementAction(I18N.gui().get("view.3d.interact"), View.Mode.INTERACT);

  private static final String FULLSCREEN_WARNING = I18N.gui().get("view.3d.fullScreenWarning");

  public ViewControlToolBar(final View view, MouseMotionHandler inputHandler) {
    super();

    this.inputHandler = inputHandler;

    setFloatable(false);

    javax.swing.ButtonGroup group = new javax.swing.ButtonGroup();

    orbitButton = new javax.swing.JToggleButton(ORBIT_ACTION);
    add(orbitButton);
    group.add(orbitButton);

    zoomButton = new javax.swing.JToggleButton(ZOOM_ACTION);
    add(zoomButton);
    group.add(zoomButton);

    moveButton = new javax.swing.JToggleButton(MOVE_ACTION);
    add(moveButton);
    group.add(moveButton);

    interactButton = new javax.swing.JToggleButton(INTERACT_ACTION);
    if (!view.viewManager.workspace.world.program().is3D) {
      add(interactButton);
      group.add(interactButton);
    }

    add(javax.swing.Box.createHorizontalStrut(8));

    status = new org.nlogo.swing.SelectableJLabel("");
    status.setFont(status.getFont().deriveFont(java.awt.Font.BOLD));
    add(status);

    add(javax.swing.Box.createHorizontalGlue());
    add(javax.swing.Box.createHorizontalStrut(8));

    final javax.swing.JButton resetButton = new javax.swing.JButton(I18N.gui().get("view.3d.resetPerspective"));
    resetButton.addActionListener(new java.awt.event.ActionListener() {
      public void actionPerformed(java.awt.event.ActionEvent e) {
        view.resetPerspective();
      }
    });
    add(resetButton);

    add(javax.swing.Box.createHorizontalStrut(8));

    final javax.swing.JButton fullscreenButton = new javax.swing.JButton(I18N.gui().get("view.3d.fullScreen"));
    fullscreenButton.addActionListener(new java.awt.event.ActionListener() {
      public void actionPerformed(java.awt.event.ActionEvent e) {
        String[] options = {I18N.gui().get("common.buttons.continue"), I18N.gui().get("common.buttons.cancel")};
        boolean isWindows = System.getProperty("os.name").toLowerCase().startsWith("win");

        if (!isWindows || view.viewManager.warned() ||
            (0 == org.nlogo.swing.OptionDialog.show
                (view, I18N.gui().get("common.messages.warning"), FULLSCREEN_WARNING,
                    options))) {
          view.viewManager.setFullscreen(true);
          view.viewManager.warned(true);
        }
      }
    });
    add(fullscreenButton);

    add(javax.swing.Box.createHorizontalStrut(16));

    orbitButton.doClick();


    add(javax.swing.Box.createHorizontalStrut(8));

    setButtonsEnabled(true);
  }

  private Perspective perspective;
  private Agent agent;

  void setStatus(Perspective perspective, Agent agent) {
    // don't update if perspective didn't change
    if (this.perspective != perspective ||
        (agent != null && !agent.equals(this.agent))) {
      this.perspective = perspective;
      this.agent = agent;
      switch (perspective) {
        case OBSERVE:
          status.setText("");
          setButtonsEnabled(true);
          break;

        case WATCH:
          status.setText(I18N.gui().get("view.3d.watching") + agent.toString());
          ORBIT_ACTION.setEnabled(true);
          ZOOM_ACTION.setEnabled(true);
          MOVE_ACTION.setEnabled(false);
          if (moveButton.isSelected()) {
            orbitButton.doClick();
          }
          break;

        case RIDE:
          status.setText(I18N.gui().get("view.3d.riding") + agent.toString());
          setButtonsEnabled(false);
          ZOOM_ACTION.setEnabled(true);
          if ((!interactButton.isSelected()) &&
              (!zoomButton.isSelected())) {
            zoomButton.doClick();
          }
          break;

        case FOLLOW:
          status.setText(I18N.gui().get("view.3d.following") + agent.toString());
          setButtonsEnabled(false);
          ZOOM_ACTION.setEnabled(true);
          if ((!interactButton.isSelected()) &&
              (!zoomButton.isSelected())) {
            zoomButton.doClick();
          }
          break;

        default:
          throw new IllegalStateException();
      }
    }
  }

  private void setButtonsEnabled(boolean enabled) {
    ORBIT_ACTION.setEnabled(enabled);
    ZOOM_ACTION.setEnabled(enabled);
    MOVE_ACTION.setEnabled(enabled);
  }

  private void setMovementMode(View.Mode mode) {
    inputHandler.setMovementMode(mode);
  }

  private final class MovementAction
      extends javax.swing.AbstractAction {
    final View.Mode mode;

    public MovementAction(String label, View.Mode mode) {
      super(label);
      this.mode = mode;
    }

    public void actionPerformed(java.awt.event.ActionEvent e) {
      setMovementMode(mode);
    }
  }
}

File: netlogo-gui/src/main/gl/view/ViewControlToolBar.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.gl.view

import java.awt.Font
import java.awt.event.ActionEvent
import javax.swing.{ AbstractAction, Box, ButtonGroup, JLabel }

import org.nlogo.api.Perspective
import org.nlogo.core.I18N
import org.nlogo.swing.{ BoxRow, Button, OptionPane, ToolBarToggleButton, Zoomable, ZoomableBorder }
import org.nlogo.theme.{ InterfaceColors, ThemeSync }

import MouseMotionHandler.{ Mode, OrbitMode, ZoomMode, TranslateMode, InterfaceMode }

class ViewControlToolBar(view: GLViewInterface, inputHandler: MouseMotionHandler) extends BoxRow(6) with ThemeSync {
  private implicit val i18nPrefix: org.nlogo.core.I18N.Prefix = I18N.Prefix("view.3d")

  private val orbitButton = new ModeButton(I18N.gui("orbit"), OrbitMode)
  private val zoomButton = new ModeButton(I18N.gui("zoom"), ZoomMode)
  private val moveButton = new ModeButton(I18N.gui("move"), TranslateMode)
  private val interactButton = new ModeButton(I18N.gui("interact"), InterfaceMode)

  private val resetButton = new Button(I18N.gui("resetPerspective"), view.resetPerspective)
  private val fullScreenButton = new Button(I18N.gui("fullScreen"), () => {
    val options = Seq(I18N.gui.get("common.buttons.continue"), I18N.gui.get("common.buttons.cancel"))
    val isWindows = System.getProperty("os.name").toLowerCase.startsWith("win")

    if (!isWindows || view.warned ||
      (new OptionPane(view, I18N.gui.get("common.messages.warning"), I18N.gui("fullScreenWarning"), options,
                      OptionPane.Icons.Warning).getSelectedIndex == 0)) {
      view.setFullscreen(true)
      view.warned = true
    }
  })

  private val status = new JLabel with Zoomable {
    setBaseFont(getFont.deriveFont(Font.BOLD))
  }

  private var perspective: Option[Perspective] = None

  locally {
    val group = new ButtonGroup

    group.add(orbitButton)
    group.add(zoomButton)
    group.add(moveButton)

    setOpaque(true)
    setBorder(new ZoomableBorder(6, 6, 6, 6))

    add(orbitButton)
    add(zoomButton)
    add(moveButton)

    if (!view.world.program.dialect.is3D) {
      add(interactButton)

      group.add(interactButton)
    }

    add(status)
    add(Box.createHorizontalGlue)
    add(resetButton)
    add(fullScreenButton)

    orbitButton.doClick()

    setButtonsEnabled(true)
  }

  def setStatus(perspective: Perspective): Unit = {
    // don't update if perspective didn't change
    if (this.perspective.orNull != perspective) {
      this.perspective = Option(perspective)

      status.setText(perspective.toString)

      perspective match {
        case Perspective.Observe =>
          setButtonsEnabled(true)

        case Perspective.Watch(a) =>
          orbitButton.setEnabled(true)
          zoomButton.setEnabled(true)
          moveButton.setEnabled(false)

          if (moveButton.isSelected)
            orbitButton.doClick()

        case Perspective.Ride(a) =>
          setButtonsEnabled(false)
          zoomButton.setEnabled(true)

          if (!interactButton.isSelected && !zoomButton.isSelected)
            zoomButton.doClick()

        case Perspective.Follow(a, _) =>
          setButtonsEnabled(false)
          zoomButton.setEnabled(true)

          if (!interactButton.isSelected && !zoomButton.isSelected)
            zoomButton.doClick()
      }
    }
  }

  private def setButtonsEnabled(enabled: Boolean): Unit = {
    orbitButton.setEnabled(enabled)
    zoomButton.setEnabled(enabled)
    moveButton.setEnabled(enabled)
  }

  override def syncTheme(): Unit = {
    setBackground(InterfaceColors.toolbarBackground())

    status.setForeground(InterfaceColors.toolbarText())

    resetButton.syncTheme()
    fullScreenButton.syncTheme()
  }

  private class ModeButton(name: String, mode: Mode)
    extends ToolBarToggleButton(new AbstractAction(name) {
      override def actionPerformed(e: ActionEvent): Unit = {
        inputHandler.setMovementMode(mode)
      }
    }) {
      setBorder(new ZoomableBorder(3, 12, 3, 12))
    }
}
